examples/embed-ingest-job-example/20_clean_and_embed_json_data.py

```diff
@@ -0,0 +1,98 @@
+import csv
+import json
+import logging
+import pathlib
+import re
+
+import nltk
+from config import DOCUMENTS_JSON_FILE_LOCATION
+from config import EMBEDDINGS_PKL_FILE_LOCATION
+from nltk.corpus import stopwords
+from nltk.stem import WordNetLemmatizer
+from sentence_transformers import SentenceTransformer
+from vdk.api.job_input import IJobInput
+
+log = logging.getLogger(__name__)
+
+
+def clean_text(text):
+    """
+    Prepares text for NLP tasks (embedding and RAG) by standardizing its form. It focuses on retaining
+    meaningful words and achieving consistency in their representation. This involves
+    converting to lowercase (uniformity), removing punctuation and stopwords
+    (focusing on relevant words), and lemmatization (reducing words to their base form).
+    Such preprocessing is crucial for effective NLP analysis.
+
+    :param text: A string containing the text to be processed.
+    :return: The processed text as a string.
+    """
+    text = text.lower()
+    # remove punctuation and special characters
+    text = re.sub(r"[^\w\s]", "", text)
+    # remove stopwords and lemmatize
+    stop_words = set(stopwords.words("english"))
+    lemmatizer = WordNetLemmatizer()
+    text = " ".join(
+        [lemmatizer.lemmatize(word) for word in text.split() if word not in stop_words]
+    )
+    return text
+
+
+def load_and_clean_documents(json_file_path):
+    cleaned_documents = []
+    with open(json_file_path, encoding="utf-8") as file:
+        documents = json.load(file)
+
+    for doc in documents:
+        if "page_content" in doc:
+            cleaned_text = clean_text(doc["page_content"])
+            cleaned_documents.append([cleaned_text])
+
+    print(len(cleaned_documents))
+    return cleaned_documents
+
+
+def embed_documents_in_batches(documents):
+    # the model card: https://huggingface.co/sentence-transformers/all-mpnet-base-v2
+    model = SentenceTransformer("all-mpnet-base-v2")
+    total = len(documents)
+    log.info(f"total: {total}")
+    embeddings = []
+    for start_index in range(0, total):
+        # the resources are not enough to batch 2 documents at a time, so the batch = 1 doc
+        batch = documents[start_index]
+        log.info(f"BATCH: {len(batch)}.")
+        embeddings.extend(model.encode(batch, show_progress_bar=True))
+
+    print(len(embeddings))
+    return embeddings
+
+
+def run(job_input: IJobInput):
+    log.info(f"Starting job step {__name__}")
+
+    data_job_dir = pathlib.Path(job_input.get_job_directory())
+    input_json = data_job_dir / DOCUMENTS_JSON_FILE_LOCATION
+    output_embeddings = data_job_dir / EMBEDDINGS_PKL_FILE_LOCATION
+
+    temp_dir = job_input.get_temporary_write_directory()
+    nltk_data_path = temp_dir / "nltk_data"
+    nltk_data_path.mkdir(exist_ok=True)
+    nltk.data.path.append(str(nltk_data_path))
+
+    nltk.download("stopwords", download_dir=str(nltk_data_path))
+    nltk.download("wordnet", download_dir=str(nltk_data_path))
```

Comment:
nit: you can move those in "prepare_nltk" or similar function

Reply:
done, moved it to setup_nltk